aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,154 @@
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", schema, ". ",
+              "We do not yet support operations with multiple returns.");
+  TORCH_WARN("Batching rule not implemented for ", schema, " falling back "
+             "to slow (for loop and stack) implementation");
+
+  const auto num_arguments = schema.arguments().size();
+  const auto arguments = torch::jit::last(stack, num_arguments);
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `arguments` they came from.
+  SmallVector<Tensor,kVmapTransformStaticInputSize> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < arguments.size(); ++idx) {
+    const auto& ivalue = arguments[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Strategy: For each batch, we are going to push slices (where applicable)
+  // of the arguments onto `stack`, call `op`, and store the result in
+  // `output_shards`.
+  std::vector<Tensor> output_shards;
+  output_shards.reserve(num_batches);
+
+  for (int64_t linear_idx = 0; linear_idx < num_batches; ++linear_idx) {
+    auto index = computeIndex(linear_idx, batch_sizes);
+    auto batched_tensor_inputs_pos_iter = batched_tensor_inputs_position.begin();
+    auto input_physical_views_iter = input_physical_views.begin();
+    for (int64_t arg_idx = 0; arg_idx < num_arguments; ++arg_idx) {
+      // The accounting for this is tricky. Assume `op` has 4 args, A, B, C, D.
+      // and `arg_idx = 2`. Every iteration of this loop pushes one element
+      // onto the stack, so we've already pushed two things (call them A' and B').
+      // At this point, the stack should look like the following:
+      // [..., A, B, C, D, A', B'] (top of stack here)
+      // To get at C (argument number 2), we have to peek `num_arguments`
+      // into the stack from the top
```

Comment:
TBH, you should just remember where the top of the stack was when you entered and then just index off of that

Reply:
(You don't want a from top of stack accessor here)

Reply:
Are you suggesting to store a pointer to the top of the stack? The problem with a pointer is that the memory location can change due to reallocations as you add new elements onto the stack due to how `std::vector` works

Reply:
(answered offline): Ed's suggestion was to store an index